### PR TITLE
Add recents to Browse models table

### LINKED ISSUE #42632: Add recents to Browse models table

### PR DESCRIPTION
**Pull request by rafpaf**
Adds a table of recently viewed models to the Browse models table

Currently only 5 models are shown, but this will change once the relevant endpoint changes (this work is currently ongoing).

image

Note that the Recents table and the table below it use different loading wrappers.

Closes #42632

### CODE REVIEW

File: frontend/src/metabase/browse/components/RecentlyViewedModels.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        blankComponent={<Box mih="129px" />}
+      >
+        <RecentlyViewedModelsGrid>
+          {recentItems.map(item => {
+            const { model_object: model, model_id } = item;
+            const modelWithId = {
+              ...model,
+              model: "dataset" as SearchModel,
+              id: model_id,
+            };
```
Comment (iethree) [inline frontend/src/metabase/browse/components/RecentlyViewedModels.tsx:41]: this weird object format will become more sane with https://github.com/metabase/metabase/issues/41957

File: frontend/src/metabase/core/components/Ellipsified/Ellipsified.tsx [hunk 2]
```diff
@@ -18,6 +18,7 @@ interface EllipsifiedProps {
   placement?: Placement;
   "data-testid"?: string;
   id?: string;
+  hideOverflowXOnly?: boolean;
```
Comment (rafpaf) [inline frontend/src/metabase/core/components/Ellipsified/Ellipsified.tsx:21]: I don't need this anymore

File: frontend/src/metabase/browse/components/ModelsTable.tsx [hunk 3]
```diff
@@ -126,9 +147,20 @@ export const ModelsTable = ({ models }: ModelsTableProps) => {
...
+                error={error}
+                loading={!!isLoading}
+              />
+            </td>
+          </LoadingAndErrorWrapperTableRow>
+        ) : (
+          sortedModels.map((model: ModelResult) => (
+            <TBodyRow model={model} key={`${model.model}-${model.id}`} />
+          ))
+        )}
```
Comment (iethree) [inline frontend/src/metabase/browse/components/ModelsTable.tsx:163]: putting the loading state inside the table is a nice UI touch, but I think it's unrelated to adding recents, so it would be best in a separate PR

File: frontend/src/metabase/browse/components/ModelsTable.tsx [hunk 4]
```diff
@@ -126,9 +147,20 @@ export const ModelsTable = ({ models }: ModelsTableProps) => {
...
       <TBody>
-        {sortedModels.map((model: ModelResult) => (
-          <TBodyRow model={model} key={`${model.model}-${model.id}`} />
-        ))}
+        {error || isLoading ? (
+          <LoadingAndErrorWrapperTableRow>
+            <td colSpan={5}>
+              <DelayedLoadingAndErrorWrapper
+                error={error}
+                loading={!!isLoading}
```
Comment (iethree) [inline frontend/src/metabase/browse/components/ModelsTable.tsx:155]: nested loading components is a bit confusing 🥴

File: frontend/src/metabase/browse/components/RecentlyViewedModels.styled.tsx [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import { Grid } from "metabase/ui";
+
+export const RecentlyViewedModelsGrid = styled(Grid)`
+  display: grid;
+  grid-template-columns: repeat(auto-fill, minmax(13rem, 1fr));
+  gap: 0.5rem;
+  margin: 0;
+  width: 100%;
+  margin-bottom: 0.5rem;
+`;
```
Comment (iethree) [inline frontend/src/metabase/browse/components/RecentlyViewedModels.styled.tsx:12]: I don't think you need to add any styling to the Grid component. In fact, I think adding CSS grid properties overwrites the flexbox grid it's going for: https://v6.mantine.dev/core/grid/

File: frontend/src/metabase/browse/components/RecentlyViewedModels.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import { getCountOfRecentlyViewedModelsToShow } from "./utils";
+
+export const RecentlyViewedModels = ({
+  modelCount,
+  applyCurrentFilters,
+}: {
+  /** The number of recently viewed models shown
+   * depends on the number of models shown in the table below. */
+  modelCount: number;
+  applyCurrentFilters: (unfilteredModels: ModelResult[]) => ModelResult[];
```
Comment (iethree) [inline frontend/src/metabase/browse/components/RecentlyViewedModels.tsx:19]: I think this gets simpler on a few levels if you do the data fetching and filtering one component up, and simply pass the already filtered models into this (now) pure component.

File: frontend/src/metabase/browse/components/BrowseModels.styled.tsx
Comment (iethree) [inline frontend/src/metabase/browse/components/BrowseModels.styled.tsx:1]: some context would be helpful here? are these deleted classes just unused styles from previous iterations of this feature?

File: frontend/src/metabase-types/entities/search.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  SearchModel,
+  SearchResult,
+} from "metabase-types/api";
+
+/** SearchResult with most properties optional */
+export type WeakSearchResult<
+  Id extends SearchResultId = SearchResultId,
+  Model extends SearchModel = SearchModel,
+> = Partial<SearchResult<Id, Model>> &
+  Pick<SearchResult<Id, Model>, "id" | "name" | "model">;
```
Comment (iethree) [inline frontend/src/metabase-types/entities/search.ts:12]: I think this should go closer to the components that use it. It might be generally useful, but it's more likely going to be used once, and will be easier to find closer to where it is actually needed

**Review by iethree — CHANGES_REQUESTED** [review]
This feels really nice. let's try to clean a few things up and simplify by:

- lifting fetching and filtering out of the recent models component
- using the existing PinnedItemCard component
- making changes to loading state/components in a separate PR

👍

File: frontend/src/metabase-types/entities/search.ts (hunk 7 above)
Comment (rafpaf) [reply frontend/src/metabase-types/entities/search.ts:12]: I originally put it in the `browse` folder, but I noticed that it needs to be used in `frontend/test/__support__/server-mocks/search.ts`. I didn't want to import something from `browse` in that file. So I moved it here

File: frontend/src/metabase/browse/components/BrowseModels.styled.tsx
Comment (rafpaf) [reply frontend/src/metabase/browse/components/BrowseModels.styled.tsx:1]: Yeah - removing dead code there

File: frontend/src/metabase/browse/components/ModelsTable.tsx (hunk 4 above)
Comment (rafpaf) [reply frontend/src/metabase/browse/components/ModelsTable.tsx:155]: I'm not seeing the nested loading components? 🤔 

`LoadingAndErrorWrapperTableRow` isn't a loading component fwiw - it's a style component that wraps the `DelayedLoadingAndErrorWrapper`

File: frontend/src/metabase/browse/components/RecentlyViewedModels.styled.tsx (hunk 5 above)
Comment (rafpaf) [reply frontend/src/metabase/browse/components/RecentlyViewedModels.styled.tsx:12]: I've removed `Grid` (its default styles don't work here)

File: frontend/src/metabase/browse/components/RecentlyViewedModels.tsx (hunk 6 above)
Comment (rafpaf) [reply frontend/src/metabase/browse/components/RecentlyViewedModels.tsx:18]: I began to do this but I ran into a complication. By moving state up, we make it a bit awkward to give the RecentlyViewedModels component its own error state.

Do you think this refactor is a blocker?

File: frontend/src/metabase/browse/components/RecentlyViewedModels.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  const headerId = "recently-viewed-models-heading";
+
+  return (
+    <>
+      <RecentlyViewedModelsHeader id={headerId} />
+      <DelayedLoadingAndErrorWrapper
+        error={error}
+        loading={isLoading}
+        blankComponent={<Box mih="129px" />}
+      >
```
[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/browse/components/RecentlyViewedModels.tsx:52]: I explored using the `if` method; I found that I would need to do:

```
if (error || loading) {
  return (<>
    <RecentlyViewedModelsHeader id={headerId} />
    <DelayedLoadingAndErrorWrapper error={error} loading={loading} />
  </>)
}

return (<>
  <RecentlyViewedModelsHeader id={headerId} />
  // etc.
</>)
```

Using the wrapper as a, well, wrapper seems cleaner. Wdyt?

File: frontend/src/metabase/browse/components/ModelCard.tsx
[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/browse/components/ModelCard.tsx:1]: I took a crack at this and found that `PinnedItemCard` is devotedly married to the `CollectionItem` type. I've adapted `ModelCard` so that it draws on prior art without any need to wrest any such art into a different shape. What do you think?

File: frontend/test/__support__/server-mocks/search.ts [hunk 9]
```diff
@@ -1,8 +1,11 @@
 import fetchMock from "fetch-mock";
 
-import type { CollectionItem, SearchResult } from "metabase-types/api";
+import type { CollectionItem } from "metabase-types/api";
+import type { WeakSearchResult } from "metabase-types/entities/search";
 
-export function setupSearchEndpoints(items: (CollectionItem | SearchResult)[]) {
+export function setupSearchEndpoints(
+  items: (CollectionItem | WeakSearchResult)[],
```
Comment (iethree) [inline frontend/test/__support__/server-mocks/search.ts:7]: I don't think we should be changing the type of the server mocks here. the server returns what it returns, the type should reflect that.

File: frontend/src/metabase/browse/components/ModelCard.tsx
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/browse/components/ModelCard.tsx:1]: Why is `PinnedItemCard` unsuitable? I think the only difference is the action menu, right?

**Review by iethree — CHANGES_REQUESTED** [review]
We should not introduce the unnecessary `ModelCard` component.
Two potential alternative approaches
a) allow the existing PinnedItemCard  to work with search payloads [example](https://github.com/metabase/metabase/pull/42661)
b) extract a data-model agnostic UI component (maybe `DetailCard`) that is used by both PinnedItemCard and ModelCard to keep the display consistent throughout the app.

**Review by iethree — APPROVED** [review]
😍  Wonderful!

**Comment by rafpaf** [comment]
Note to self: need to fix e2e test in browse.cy

**Comment by rafpaf** [comment]
### Merge activity

* **Jun 7, 4:28 PM EDT**: @rafpaf merged this pull request with [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/42138).
